polar-core/src/normalize.rs

```diff
@@ -0,0 +1,317 @@
+            Err(_) => self.clone(),
+            Ok(Operation {
+                operator: Not,
+                args,
+            }) => args[0].nnf().negated(),
+            Ok(Operation { operator, args }) => {
+                self.clone_with_value(Value::Expression(Operation {
+                    operator: *operator,
+                    args: args.iter().cloned().map(|t| t.nnf()).collect(),
+                }))
+            }
+        }
+    }
+
+    fn _hs(&self, n: usize) -> Self {
+        self.value().as_expression().unwrap().args[n].clone()
+    }
+
+    fn lhs(&self) -> Self {
+        self._hs(0)
+    }
+    fn rhs(&self) -> Self {
+        self._hs(1)
+    }
+
+    /// normalize an expression by fully distributing logical connectives. the
+    /// expression must already be in btnf and nnf. the arguments are two
+    /// predicate / constructor pairs:
+    /// - p1/c1: predicate/constructor for *inner* connective (and for dnf, or for cnf)
+    /// - p2/c2: predicate/constructor for *outer* connective (or for dnf, and for cnf)
+    fn distribute(
+        &self,
+        p1: fn(&Term) -> bool,
+        c1: fn(Term, Term) -> Term,
+        p2: fn(&Term) -> bool,
+        c2: fn(Term, Term) -> Term,
+    ) -> Self {
+        use Value::*;
+
+        match self.value().as_expression() {
+            Err(_) => self.clone(),
+            Ok(Operation { operator, args }) => {
+                let args: Vec<_> = args.iter().map(|t| t.distribute(p1, c1, p2, c2)).collect();
+
+                if p1(self) && p2(&args[0]) {
+                    c2(
+                        c1(args[0].lhs(), args[1].clone()),
+                        c1(args[0].rhs(), args[1].clone()),
+                    )
+                    .distribute(p1, c1, p2, c2)
+                } else if p1(self) && p2(&args[1]) {
+                    c2(
+                        c1(args[0].clone(), args[1].lhs()),
+                        c1(args[0].clone(), args[1].rhs()),
+                    )
+                    .distribute(p1, c1, p2, c2)
+                } else {
+                    self.clone_with_value(Expression(Operation {
+                        operator: *operator,
+                        args,
+                    }))
+                }
+            }
+        }
+    }
+
+    /// Condition input for dnf/cnf transformation
+    /// - negations fully nested
+    /// - double negatives removed
+    /// - and/or nodes form a binary tree
+    fn pre_norm(&self) -> Self {
+        self.btnf().nnf()
+    }
+
+    pub fn dnf(&self) -> Self {
+        self.pre_norm().distribute(andp, and_, orp, or_)
+    }
+
+    pub fn cnf(&self) -> Self {
+        self.pre_norm().distribute(orp, or_, andp, and_)
+    }
+}
+
+pub fn or_(l: Term, r: Term) -> Term {
+    term!(op!(Or, l, r))
+}
+
+pub fn and_(l: Term, r: Term) -> Term {
+    term!(op!(And, l, r))
+}
+
+pub fn not_(t: Term) -> Term {
+    term!(op!(Not, t))
+}
+
+fn opnp(l: &Term, op: Operator) -> bool {
```

Comment:
What does this mean? `is_op` seems like an easier to understand and similarly short name.

Reply:
Now realizing `p` means predicate. In any case, this is a reason why we should use descriptive names.

Reply:
changed